README.md
# peony-extensions

## Getting Start
- Make sure all build dependencies have been installed in your machine. You can reference debian/control for building depends.

* Create *build* directory and change to it, then build with cmake. The follow
  commands will auto build extensions as library.

``` sh
#cmake ..
qmake ..
make
```
- Put the generated libraries into &&[QT_INSTALL_LIBS]/peony-extensions.
  
- Test the extension with peony.

## Current extensions

- **terminal**, open terminal with right click menu.
- **shared**, a share properties page in properties window.
- ~~**admin**, open a directory or a file as root with menu.~~
- **parchives**, compress/uncompress files with menu.
- **computer view**, a special view for displaying computer:///.
- **send-to-device**, provides a directory menu action for sending selected files to mounted device.
- **bluetooth**, provides a directory menu action for sneding selected files by bluetooth, depending on ukui-bluetooth.
- **set-wallparper**, privides a directory menu action for set desktop background.

## binary compatibility
Peony-Extensions strongly relies on Peony library, that means the binary files of those plugins might not compat with Peony when the library updated and some API changed. Sometimes the old plugins will not be loaded, or will make Peony crashed due to call an unmatch method. To avoid the binary compatibility problems, we add a version checkment in latest version both in this project and Peony. The plugins' version infomation are loaded from [common.json](common.json), this must match to the Peony VERSION marco.

When peony extensions doesn't work or can't be build correctly, please consider using latest Peony library built locally, and rebuild this project again.
## Translations
Unlike Peony, Peony Extensions translations job is relatively primitive and crude. The main idea is every plugin translates itself, and compile translations resources with code (qrc).

When you are going to translate peony extensions (I created), here are the mainly steps.

- use lupdate to generate .ts file for your language, the naming rules reference existing files.
- use lrelease to publish .ts file to .qm file.
- modify the project's .qrc file, add your .qm file into recources to be compiled.


debian/changelog
peony-extensions (3.14.1.0) v101; urgency=medium

  * Bug号: 无
  * 需求号:
    #13398:增加文件夹共享时增加共享徽标（向文杰）
  * 其它改动说明: 无
  * 影响域:
    共享文件夹&角标显示

 -- Yue Lan <lanyue@kylinos.cn>  Mon, 09 May 2022 16:15:47 +0800

peony-extensions (3.14.0+0426) v101; urgency=medium

  * Bug号: 无
  * 需求号: 无
  * 其它改动说明:
    #110317 #110543 【panguw】 解决计算机视图数据光盘挂载后状态未刷新问题（杨玲）
    【主线自测】修复有些分区无法重命名的情况（向文杰）
    #104728 修复盘古w计算机视图的数据盘卸载后还是显示DATA的问题（向文杰）
    #107615 【panguw】修复计算机视图卸载和挂载时没有系统提示（向文杰，通用）
  * 影响域:
    计算机视图&卸载/挂载操作、分区重命名

 -- Yue Lan <lanyue@kylinos.cn>  Tue, 26 Apr 2022 12:20:39 +0800

peony-extensions (3.2.3+0408) v101; urgency=medium

  * sp1 2203版本更新
    - Bug号: 无
    - 需求号: 无
    - 其它改动说明:
      #110317 #110543 【panguw】 解决计算机视图数据光盘挂载后状态未刷新问题（杨玲）
      【主线自测】修复有些分区无法重命名的情况（向文杰）
    - 影响域:
      计算机视图、分区重命名

 -- Yue Lan <lanyue@kylinos.cn>  Fri, 08 Apr 2022 15:15:15 +0800

peony-extensions (3.2.3+0318) v101; urgency=medium

  * sp1 2203版本更新
    - Bug号:
      无
    - 需求号:
      无
    - 其它改动说明:
      #104728 修复盘古w计算机视图的数据盘卸载后还是显示DATA的问题（向文杰）
      #107615 【panguw】修复计算机视图卸载和挂载时没有系统提示（向文杰，通用）
    - 影响域:
      计算机视图&卸载/挂载操作

 -- Yue Lan <lanyue@kylinos.cn>  Fri, 18 Mar 2022 11:28:22 +0800

peony-extensions (3.2.3+0218) v101; urgency=medium

  * sp2 2201-5版本更新
    - Bug号:
      #63375 修复在计算机视图右键双击偶现进入系统的问题（向文杰）
    - 需求号: 无
    - 其它改动说明:
      #105070 为计算机视图增加重命名功能（兰悦 panguw）
      #55297,102968 右键点击分卷压缩文件没有”解压缩到此处“的功能（何美宏 OEM）
      #105475 修复不同格式的U盘重命名时失败提示信息不准确的问题（向文杰 panguw）
      #104728 修复盘古w计算机视图的数据盘未汉化的问题（向文杰 panguw）
    - 影响域:
      U盘重命名、计算机视图、压缩、数据盘显示

 -- Yue Lan <lanyue@kylinos.cn>  Fri, 18 Feb 2022 16:49:05 +0800

peony-extensions (3.2.3+0121) v101; urgency=medium

  * sp2 2201-4版本更新
    - Bug号:
      #101846 【文件管理器】当字号为15时，计算机-网上邻居里名称过长的主机，其第四行显示不全，有截断（向文杰）
      #99662 【文件管理器】依次拖动本地分区文件系统和数据盘图标，再选中文件系统图标右键查看属性无反应（向文杰）
    - 需求号: 无
    - 其它改动说明:
      #102556 samba关闭网络，文件管理器右键菜单卡顿（丁敬）
    - 影响域:
      计算机视图、发送到移动设备

 -- Yue Lan <lanyue@kylinos.cn>  Fri, 21 Jan 2022 11:56:54 +0800

peony-extensions (3.2.3+0112) v101; urgency=medium

  * sp2 2201-4版本更新
    - Bug号:
      #92671 【文件管理器】接入手机后，在左侧边栏对其右键的右键菜单(卸载/属性)，与在计算机右侧界面对其右键的右键菜单(卸载/弹出/格式化/属性)，内容不一致（杨玲）
    - 需求号: 无
    - 其它改动说明:
      仅版本升级
    - 影响域:
      手机卸载弹出操作菜单

 -- Yue Lan <lanyue@kylinos.cn>  Wed, 12 Jan 2022 14:21:39 +0800

peony-extensions (3.2.3+0107) v101; urgency=medium

  * sp2 2201-4版本更新
    - Bug号: 无
    - 需求号: 无
    - 其它改动说明:
      仅版本升级
    - 影响域:
      无

 -- Yue Lan <lanyue@kylinos.cn>  Fri, 07 Jan 2022 16:26:09 +0800

peony-extensions (3.2.2+1231) v101; urgency=medium

  * sp2 2201-3版本更新
    - Bug号:
      #91441 【用例204892】【需求7682】对于镜像盘依然会有重命名选项，重命名后又会提示设备不支持，重命名失败（丁敬）
      #99277 修复容量比例较大时，进度条末端样式不统一（向文杰）
    - 需求号: 无
    - 其它改动说明:
      修复蓝牙接口不识别英文括号的问题（李鑫）
      修改debian/control修复自适应升级问题（何美宏）
    - 影响域:
      升级测试、镜像盘重命名、计算机视图样式、发送到蓝牙

 -- Yue Lan <lanyue@kylinos.cn>  Fri, 31 Dec 2021 13:51:08 +0800

peony-extensions (3.2.2+1227) v101; urgency=medium

  * sp2 2201-3版本更新
    - Bug号:
      #99277 【SP2 UI走查】【文件管理器】计算机首页 - 进度条样式未统一（向文杰）
      #99284 【SP2 UI走查】【文件管理器】计算机首页 - “本地分区”和“网络邻居”需调整（向文杰）
      #99295 【SP2 UI走查】【文件管理器】计算机首页 - 页面上存在过多无用功能（向文杰）
    - 需求号: 无
    - 其它改动说明: 无
    - 影响域:
      计算机视图

 -- Yue Lan <lanyue@kylinos.cn>  Mon, 27 Dec 2021 14:32:04 +0800

peony-extensions (3.2.2+1220) v101; urgency=medium

  * sp2 1220版本更新
    - Bug号:
      #98496 修复U盘重命名时，弹窗未汉化的问题（向文杰）
    - 需求号: 无
    - 其它改动说明:
      修复容量字体对应的主题色板值（何美宏）
    - 影响域:
      U盘、设备重命名

 -- Yue Lan <lanyue@kylinos.cn>  Mon, 20 Dec 2021 16:45:30 +0800

peony-extensions (3.2.2+1215) v101; urgency=medium

  * sp2 1220版本更新
    - Bug号:
      #96652 修复移动硬盘分区后小于128G显示的是U盘图标问题（向文杰）
      #97478 修复翻译信息不准确的问题（向文杰）
    - 需求号: 无
    - 其它改动说明: 无
    - 影响域:
      U盘、设备重命名

 -- Yue Lan <lanyue@kylinos.cn>  Wed, 15 Dec 2021 17:20:08 +0800

peony-extensions (3.2.2+1210) v101; urgency=medium

  * sp2 1220版本更新
    - Bug号:
      #91441 对于镜像盘依然会有重命名选项，重命名后又会提示设备不支持，重命名失败（丁敬）
      #93280 修复U盘重命名时，首字符带有.重命名后无法挂载问题（向文杰）
      #92731 【文件浏览器】普通用户弹出管理器用户的U盘时，无任何弹窗，U盘也未被弹出（期望：弹出授权窗口）（兰悦）
    - 需求号: 无
    - 其它改动说明: 无
    - 影响域:
      设备重命名、弹出卸载操作

 -- Yue Lan <lanyue@kylinos.cn>  Fri, 10 Dec 2021 15:56:36 +0800

peony-extensions (3.2.2+1129) v101; urgency=medium

  * sp2 1122版本
    - Bug号:
      #90641:接入type-C拓展坞后，文件浏览器的计算机右侧界面或左侧边栏计算机下会显示两个Generic MassStorageClass（何美宏）
    - 需求号: 无
    - 其它改动说明: 无
    - 影响域:
      计算机视图设备显示

 -- Yue Lan <lanyue@kylinos.cn>  Mon, 29 Nov 2021 14:58:50 +0800

peony-extensions (3.2.2+1124) v101; urgency=medium

  * sp2 1122版本
    - Bug号:
      #82067 【文件浏览器】从侧边栏计算机下右键手机-卸载，计算机右侧界面的手机下仍显示容量数据和使用进度条（期望：显示“需要首先挂载分区”）（杨玲）
      #90042 计算机右侧界面的共享数据，其预览界面文件名称是“本机共享”、其右键属性界面也显示为“本机共享"（何文飞）
      #89255 计算机视图共享数据歧义问题 （兰悦）
      #89975 【文件浏览器】当计算机右侧界面的共享数据显示占用空间很大(如94.1G/99.2G)，但点进去显示空白（兰悦）
    - 需求号: 无
    - 其它改动说明:
      自动分区相关问题重新按照最新结论进行改动，计算机视图的共享数据改为数据盘指向/data，侧边栏增加数据盘
    - 影响域:
      计算机视图&自动分区，手机&卸载，本机共享&属性窗口

 -- Yue Lan <lanyue@kylinos.cn>  Wed, 24 Nov 2021 17:20:57 +0800

peony-extensions (3.2.2+1111) v101; urgency=medium

  * sp2 1108 版本
    - Bug号: 无
    - 需求号: 无
    - 其它改动说明:
      #00000 同步bluetooth右键发送文件的dbus接口（何美宏）
      #57473 发送到移动设备屏蔽只读U盘（丁敬）
    - 影响域:
      蓝牙、发送到移动设备

 -- Yue Lan <lanyue@kylinos.cn>  Thu, 11 Nov 2021 09:32:39 +0800

peony-extensions (3.2.2+1105) v101; urgency=medium

  * sp2 1108 版本
    - Bug号: 无
    - 需求号:
      #7682 task#58635 【文件管理器】U盘重命名（丁敬）

 -- Yue Lan <lanyue@kylinos.cn>  Fri, 05 Nov 2021 17:28:51 +0800

peony-extensions (3.2.2+1104) v101; urgency=medium

  * sp2 1108 版本
    - Bug号:
      #82320 【文件浏览器】计算机界面共享数据的详细信息里，文件名称是usershare英文；共享数据的右键属性界面，左上角也是"usershare属性"（期望：都是“共享数据”中文）（何文飞）
    - 需求号: 无
    - 其它改动说明:
      修改usershare目录的显示名称，更新翻译文件
    - 影响域:
      usershare目录

 -- Yue Lan <lanyue@kylinos.cn>  Thu, 04 Nov 2021 09:52:36 +0800

peony-extensions (3.2.2+1028) v101; urgency=medium

  * sp2 1108 版本（需首先构建完成peony 1028）
    - Bug号:
      #83206 U盘既有卸载又有弹出选项（杨玲）
    - 需求号: 无
    - 其它改动说明: 无
    - 影响域:

peony-drive-rename/drive-rename.cpp
/*
 * Peony-Qt's Library
 *
 * Copyright (C) 2021, KylinSoft Co., Ltd.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library.  If not, see <https://www.gnu.org/licenses/>.
 *
 * Authors: deng hao <denghao@kylinos.cn>
 *
 */

#include "drive-rename.h"

#include <QDebug>
#include <file-info.h>
#include <QMessageBox>
#include <QInputDialog>
#include <gio/gio.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <sys/stat.h>
#include <udisks/udisks.h>
#include <QTranslator>
#include <QFile>
#include <QApplication>
#include <file-utils.h>

typedef struct _DeviceRenameData        DeviceRenameData;

int device_rename(const char* devName, const char* name);
static void udisk_umounted (GMount* mount, GAsyncResult *res, gpointer udata);
UDisksObject* getObjectFromBlockDevice(UDisksClient* client, const gchar* bdevice);

struct _DeviceRenameData
{
    QString             devName;
    QString             rename;
    Peony::DriveRename* pThis;
};

Peony::DriveRename::DriveRename(QObject *parent) : QObject(parent), mEnable(true)
{
    QTranslator *t = new QTranslator(this);
    t->load(":/translations/peony-drive-rename_"+QLocale::system().name());
    QApplication::installTranslator(t);
}

QList<QAction *> Peony::DriveRename::menuActions(Peony::MenuPluginInterface::Types types, const QString &uri, const QStringList &selectionUris)
{
    QList<QAction *> l;

    if (selectionUris.count() != 1 || types != Peony::MenuPluginInterface::Type::SideBar) {
        return l;
    }

    QString suri = selectionUris.first();

    g_autoptr(GFile) file = g_file_new_for_uri(suri.toUtf8().constData());
    g_return_val_if_fail(file, l);

    // check is mount point and get dev name
    g_autoptr (GError) error = NULL;
    g_autoptr (GFileInfo) fileInfo = g_file_query_info (file, G_FILE_ATTRIBUTE_UNIX_IS_MOUNTPOINT "," G_FILE_ATTRIBUTE_MOUNTABLE_CAN_UNMOUNT "," G_FILE_ATTRIBUTE_MOUNTABLE_UNIX_DEVICE_FILE "," G_FILE_ATTRIBUTE_STANDARD_TARGET_URI, G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS, NULL, &error);
    if (error) qDebug() << error->message;
    g_return_val_if_fail (G_IS_FILE_INFO (fileInfo) && !error, l);

    gboolean canUmount = g_file_info_get_attribute_boolean (fileInfo, G_FILE_ATTRIBUTE_MOUNTABLE_CAN_UNMOUNT);
    gboolean isMountPoint = g_file_info_get_attribute_boolean (fileInfo, G_FILE_ATTRIBUTE_UNIX_IS_MOUNTPOINT);

    g_autofree char* devName = g_file_info_get_attribute_as_string (fileInfo, G_FILE_ATTRIBUTE_MOUNTABLE_UNIX_DEVICE_FILE);

    bool canRename = (isMountPoint || canUmount || devName);

    qDebug() << "uri: " << uri << "  " << (canRename ? "can rename" : "can not rename");
    g_return_val_if_fail (canRename, l);

    // if device has mounted
    GMount* mount = NULL;
    if (isMountPoint || canUmount) {
        if (uri.startsWith ("computer://")) {
            g_autofree char* targetUri = g_file_info_get_attribute_as_string (fileInfo, G_FILE_ATTRIBUTE_STANDARD_TARGET_URI);
            if (targetUri) {

                // related to: #105070
//                bool isWayland = qApp->property("isWayland").toBool();
//                if (!isWayland) {
//                    QString tmp = targetUri;
//                    if (tmp == "file:///data") {
//                        // 如果是data盘，则跳过
//                        return l;
//                    }
//                }

                bool isWayland = qApp->property("isWayland").toBool();
                if (isWayland) {
                    QString tmp = targetUri;
                    if (tmp == "file:///data") {
                        return l;
                    }
                }

                g_autoptr (GFile) tfile = g_file_new_for_uri (targetUri);
                if (G_IS_FILE (tfile)) {
                    mount = g_file_find_enclosing_mount (tfile, NULL, &error);
                    if (error) qDebug() << error->message;
                }
            }
        } else {
            mount = g_file_find_enclosing_mount (file, NULL, &error);
            if (error) qDebug() << error->message;
        }

        g_return_val_if_fail (G_IS_MOUNT (mount), l);

        QString type = Peony::FileUtils::getFileSystemType(suri);
        if(type.startsWith("iso")){
            g_autoptr (GFile) root = g_mount_get_root (mount);
            g_autoptr (GFileInfo) rootInfo = g_file_query_info (root, "access::", G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS, NULL, &error);
            if (error)      qDebug() << "error: " << error->message;
            bool canRename = true;
            if (G_IS_FILE_INFO (rootInfo)) {
                bool write = true;
                if (g_file_info_has_attribute (rootInfo, G_FILE_ATTRIBUTE_ACCESS_CAN_WRITE)) {
                    write = g_file_info_get_attribute_boolean (rootInfo, G_FILE_ATTRIBUTE_ACCESS_CAN_WRITE);
                }
                bool execute = true;
                if (g_file_info_has_attribute (rootInfo, G_FILE_ATTRIBUTE_ACCESS_CAN_EXECUTE)) {
                    execute = g_file_info_get_attribute_boolean (rootInfo, G_FILE_ATTRIBUTE_ACCESS_CAN_EXECUTE);
                }

                if (!write || !execute)   canRename = false;
            }

            g_return_val_if_fail (canRename, l);
        }

        if (!devName) {
            g_autoptr (GVolume) volume = g_mount_get_volume (mount);
            g_return_val_if_fail (G_IS_VOLUME (volume), l);
            devName = g_volume_get_identifier (volume, G_DRIVE_IDENTIFIER_KIND_UNIX_DEVICE);
        }
    }


    g_return_val_if_fail (devName, l);

    // 注意要筛选的设备，块设备、非根设备
    // 判断是否为光盘设备(光盘设备不允许重命名)
    if (suri.startsWith("computer:///") && suri.endsWith(".drive")
            && 0 != g_ascii_strncasecmp ("/dev/cd", devName, 7)
            && 0 != g_ascii_strncasecmp ("/dev/sr", devName, 7)
            && 0 != g_ascii_strncasecmp ("/dev/loop", devName, 9)) {
        mDevName = devName;
        QAction* action = new QAction(tr("Rename"));
        if (action) {
            l << action;
        }

        // 这里 action 弹框并执行重命名函数
        connect(action, &QAction::triggered, this, [=] () {

            // 输入框出来，提示用户U盘重命名为 ...
            // 获取到用户的输入
            // 执行重命名
            bool ok = false;
            QString text = QInputDialog::getText(nullptr, tr("Rename"), tr("Device name:"), QLineEdit::Normal, "", &ok);
            if (ok && !text.isNull() && !text.isEmpty()) {
                //首字符是.提示非法 bug#93280
                if(text.at(0) == '.'){
                    QMessageBox::warning(nullptr, tr("Warning"), tr("Renaming cannot start with a decimal point, Please re-enter!"), QMessageBox::Ok);
                    return;
                }

                QString type = Peony::FileUtils::getFileSystemType(suri);
                if((type.compare(QString::fromLocal8Bit("vfat")) == 0 && text.toUtf8().length() > 11)
                        || (type.compare(QString::fromLocal8Bit("exfat")) == 0 && text.toUtf8().length() > 15)
                        || (type.compare(QString::fromLocal8Bit("ext4")) == 0 && text.toUtf8().length() > 16)){
                    QMessageBox::warning(nullptr, tr("Warning"), tr("The device name exceeds the character limit, rename failed!"), QMessageBox::Ok);
                    return;
                }

                // 修改名字
                if (mount) {
                    int ret = QMessageBox::warning (nullptr, tr("Warning"), tr("Renaming will unmount the device. Do you want to continue?"), QMessageBox::Ok | QMessageBox::Cancel);
                    if (QMessageBox::Cancel == ret) return;

                    // 释放、释放 GMount
                    DeviceRenameData* data = new DeviceRenameData;
                    data->devName = mDevName;
                    data->rename = text;
                    data->pThis = this;
                    g_autoptr (GMountOperation) mount_op = g_mount_operation_new();
                    g_mount_unmount_with_operation (mount, G_MOUNT_UNMOUNT_NONE, mount_op, NULL, (GAsyncReadyCallback) udisk_umounted, data);
                } else {
                    int ret = device_rename(mDevName.toUtf8().constData(), text.toUtf8().constData());
                    if (0 != ret) {
                        QMessageBox::warning (nullptr, tr("Warning"), tr("The device may not support the rename operation, rename failed!"), QMessageBox::Ok);
                    }
                }
            }
        });
    }

    Q_UNUSED(uri)
    Q_UNUSED(types)

    return l;
}


int device_rename(const char* devName, const char* name)
{
    //判断参数个数是否合法
    g_return_val_if_fail(devName && name, -1);

    UDisksClient* client = udisks_client_new_sync(NULL, NULL);
    g_return_val_if_fail(client, -1);


    UDisksObject* udiskObj = getObjectFromBlockDevice(client, devName);
    g_return_val_if_fail(udiskObj, -1);

    //从设备名获取文件系统类型
    UDisksFilesystem* diskFilesystem = udisks_object_get_filesystem(udiskObj);
    g_return_val_if_fail(diskFilesystem, -1);
    GVariantBuilder optionsBuilder;
    g_variant_builder_init(&optionsBuilder, G_VARIANT_TYPE_VARDICT);
    g_variant_builder_add (&optionsBuilder, "{sv}", "label", g_variant_new_string (devName));
    g_variant_builder_add (&optionsBuilder, "{sv}", "take-ownership", g_variant_new_boolean (TRUE));

    g_autoptr (GError) error = NULL;
    gboolean ret = udisks_filesystem_call_set_label_sync (diskFilesystem, name, g_variant_builder_end(&optionsBuilder), NULL, &error);
    if (error) {
        qDebug() << error->message;
    }

    return ret ? 0 : -1;
}


UDisksObject* getObjectFromBlockDevice(UDisksClient* client, const gchar* bdevice)
{
    struct stat statbuf;
    UDisksBlock* block = NULL;
    UDisksObject* object = NULL;
    UDisksObject* cryptoBackingObject = NULL;
    g_autofree const gchar* cryptoBackingDevice = NULL;

    g_return_val_if_fail(stat(bdevice, &statbuf) == 0, object);

    block = udisks_client_get_block_for_dev (client, statbuf.st_rdev);
    g_return_val_if_fail(block != NULL, object);

    object = UDISKS_OBJECT (g_dbus_interface_dup_object (G_DBUS_INTERFACE (block)));

    cryptoBackingDevice = udisks_block_get_crypto_backing_device ((udisks_object_peek_block (object)));
    cryptoBackingObject = udisks_client_get_object (client, cryptoBackingDevice);
    if (cryptoBackingObject != NULL) {
        g_object_unref (object);
        object = cryptoBackingObject;
    }

    g_object_unref (block);

    return object;
}

static void udisk_umounted (GMount* mount, GAsyncResult *res, gpointer udata)
{
    g_autoptr (GError) error = NULL;

    int ret = -1;
    DeviceRenameData* data = (DeviceRenameData*) udata;

    if (g_mount_unmount_with_operation_finish (G_MOUNT (mount), res, &error)) {
        ret = device_rename (data->devName.toUtf8 ().constData (), data->rename.toUtf8 ().constData ());
        error = g_error_new (1, G_IO_ERROR_FAILED, data->pThis->tr("The device may not support the rename operation, rename failed!").toUtf8 ().constData (), NULL);
    } else {
        if (error) qDebug() << error->message;
    }

    if (ret != 0 && error) {
        QMessageBox::warning(nullptr, data->pThis->tr("Warning"), error->message, QMessageBox::Ok);
    }

    if (data)   delete data;
    if (mount)  g_object_unref (mount);
}
